[WIP] Add a page with rapids deployment info (#46)
rapidsai/docs: deployment/deployment.md

---
layout: default
title: Deploying RAPIDS
nav_order: 5
permalink: deployment
---

# Deploying RAPIDS
{: .no_toc}

> The RAPIDS collections of libraries can be deployed across multiple environments ranging from local nodes with a single GPU to complex HPC and cloud clusters. Highlighted below are some methods to deploy rapids libraries in different environments.

Author: - [ ] capitalize RAPIDS , and maybe Cloud

{: .fs-6 .fw-300}

### Single Node Single GPU
Running RAPIDS libraries in a Single Node Single GPU environment does not involve any additional setup. [Install](https://rapids.ai/start.html#get-rapids) the libraries you need and run in your preferred environment.
```
import cudf, cuml, cugraph, cuspatial # Good to go!
```
<!--
Modify the codeblock if needed
-->
{: .mb-7 }

### Single Node Multi GPU
RAPIDS libraries have the ability to scale up and take advantage of multiple GPUs on a single node. The [Dask-CUDA](https://github.com/rapidsai/dask-cuda) library makes it simple to set up workers on a single node with the `LocalCUDACluster` method.
```

> from dask_cuda import LocalCUDACluster

Author: All python blocks prefixed with `>>>`

from dask.distributed import Client
import dask_cudf

IPADDR = "0.0.0.0" # Update to use IP of the node
cluster = LocalCUDACluster(ip= IPADDR)
client = Client(clsuter)
```

> This [notebook](https://github.com/rapidsai/notebooks-contrib/blob/master/intermediate_notebooks/examples/weather.ipynb) demonstrates the approach discussed above.

Reviewer: I don't think we really need to refer to this example here. Not sure it adds anything

{: .mb-7 }
<!--

> Include output of client maybe?

Reviewer: Good idea. Perhaps a screenshot of the Jupyter rendering in case you can't get text to render well in this webpage

Choose a beginner example instead of intermediate?
-->

> ### Multi Node Multi GPU

Author: How much of this is cuDF specific only and not RAPIDS in general? I know other libraries like cuML might be able to do SNMG in a different way using NCCL maybe?

Reviewer: @robocopnixon can you comment here? Is there a different set of setup steps for any aspects of multi GPU or multi-node cuML?

Reviewer 2: These steps work for cuML. Some older stuff did not use Dask, but the new stuff like RF and K-Means does.

> There are multiple methods of deploying RAPIDS in a Multi-Node setting.

Reviewer: ```suggestion
```

#### **Manual Setup**

> This is the most fundamental way to deploy RAPIDS on multiple machines.

Reviewer: ```suggestion
This is the most basic means of deploying RAPIDS to multiple hosts. Most more advanced deployments use the same commands with environment configuration & parameter variations.
```

- Launch the `dask-scheduler` on one node
    ```
    $ dask-scheduler
    ```

> - Launch `dask-cuda-worker` on the rest of the nodes providing the address to the node hosting the scheduler

Reviewer: ```suggestion
- Run `dask-cuda-worker` on each worker node, providing the scheduler node's address. `dask-cuda-worker` will detect available GPUs and fork, starting one dask worker process per GPU.
```

    ```
    $ dask-cuda-worker 127.0.0.1:8786
    Start worker at:   tcp://127.0.0.1:36124
    Registered to: tcp://127.0.0.1:8786
    ```

> - Connect to the cluster in python

Reviewer: ```suggestion
- Connect to the cluster from a Python script:
```

    ```
    from dask.distributed import Client

>     client = Client(address="127.0.0.1:8786")

Reviewer: ```suggestion
# wrap Dask client API use in a main check
if __name__ == "__main__":
    client = Client(address="127.0.0.1:8786")
    use_dask_APIs()
```

    ```
    Notes:

>     - If the node has access to multiple GPUs the `dask-cuda-worker` CLI will start one worker per GPU on the node.

Reviewer: ```suggestion
```

>     - For more information on the CLI options use `dask-scheduler --help` and `dask-cuda-worker --help`

Reviewer: ```suggestion
    - For more information on Dask CLI options use `dask-scheduler --help` and `dask-cuda-worker --help`
```

    More information on deploying manually can be found [here](https://docs.dask.org/en/latest/setup/cli.html).

>     Dask also supports other options like [SSH](https://docs.dask.org/en/latest/setup/ssh.html) and [NFS](https://docs.dask.org/en/latest/setup/hpc.html#using-a-shared-network-file-system-and-a-job-scheduler).

Reviewer: ```suggestion
```

I believe @jacobtomlinson is updating dask-cuda to support SSH deployments.. let's wait for those updates before mentioning here. I'd rather provide a working example than just linking and leaving it to the user to figure out.

Reviewer 3: It is already possible to use `dask-cuda-worker` with the SSH cluster manager. The change I made this week was to add this support into the experimental rewrite of the SSH cluster manager as that class will likely be used as the foundation for the new DGX cluster manager that @pentschev is working on.

```console
# Works today
$ dask-ssh 10.0.0.1 10.0.0.2 10.0.0.3 10.0.0.4 --remote-dask-worker dask_cuda.dask_cuda_worker
```

>     While not tested more configuration options to deploy dask in a distributed setting can be found [here](https://docs.dask.org/en/latest/setup.html). 

Author: - tested`,`
- Dask

Should we link to these options as well (lost of dask-jobqueue) and other configs that I'm not sure have been tested

>     While not tested more configuration options to deploy dask in a distributed setting can be found [here](https://docs.dask.org/en/latest/setup.html). 

Reviewer: ```suggestion
```

>     Note that in the context of RAPIDS instead of the tradititonal `dask-worker` we recommend using the `dask-cuda-worker` CLI.

Reviewer: ```suggestion
    `dask-cuda-worker` is a convenient and optional extension of the normal `dask-worker` CLI command. You can also use `dask-worker` directly, manually configuring which GPUs are accessible to which workers:
```
CUDA_VISIBLE_DEVICES=0 dask-worker localhost:8786 --nprocs 1 --nthreads 1 --memory-limit 0
CUDA_VISIBLE_DEVICES=1 dask-worker localhost:8786 --nprocs 1 --nthreads 1 --memory-limit 0
```

{: .mb-5}

> #### **Kubernetes and Helm**

Author: This section definitely needs review and verification. 

cc: @randerzander and @millerhooks

> RAPIDS libraries can also be deployed using [Kubernetes](https://kubernetes.io) and [Helm](https://helm.sh). This is useful for deploying RAPIDS on Cloud services like [AWS](https://aws.amazon.com), [GCP](https://cloud.google.com) or [Azure](https://azure.microsoft.com/en-us/).

Reviewer: ```suggestion
RAPIDS libraries can also be deployed using [Kubernetes](https://kubernetes.io) and [Helm](https://helm.sh). This is useful for deploying RAPIDS on existing Kubernetes clusters, or to managed cloud Kubernetes services.
```

> - Install Helm based on the [docs](https://helm.sh/docs/using_helm/#installing-helm)

Reviewer: ```suggestion
- [Install Helm](https://helm.sh/docs/using_helm/#installing-helm)
```

> - Setup a cluster in Google Kubernetes Engines (optional)

Reviewer: ```suggestion
```

- [Link-to-k8-script-and-charts]
- ```
    $ helm install [link-to-chart] --name rapids-dask --namespace rapids
  ```
- Check the status of the running pods
  ```
  kubectl get pods -w --namespace rapids
  ```

> For more information on installing with Kubernetes and Helm visit [link-to-k8-repo]

Reviewer: ```suggestion
Additional Information:
[GPU scheduling in Kubernetes](https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/) 
[Dask Kubernetes](https://kubernetes.dask.org/en/latest/) 
```

{: .mb-5}

#### **Docker**

> ##### **Docker Compose**

Author: Would love input for these sections.
@randerzander

- Get the docker yml file [link-to-repo]
- ```
  $ docker compose ...
  ```

##### **Docker Swarm**

#### **Cloud deployments**

> For deploying RAPIDS on common cloud providers we recommend using Kuberentes and Helm. You can also deploy RAPIDS in other environments.

Author: Do we say `recommend K8's and Helm for most cloud deployments`

Reviewer: ```suggestion
You can use a single type of configuration to deploy RAPIDS across multiple cloud providers with Kubernetes and Helm. You can also deploy RAPIDS through other managed service offerings where available.
```

> For deploying RAPIDS on common cloud providers we recommend using Kuberentes and Helm. You can also deploy RAPIDS in other environments.

Reviewer: ```suggestion
```

Choose a beginner example instead of intermediate?
-->
### Multi Node Multi GPU

#### **Manual Setup**
This is the most basic means of deploying RAPIDS to multiple hosts. Most more advanced deployments use the same commands with environment configuration & parameter variations.
- Run the `dask-scheduler` on one node
    ```
    $ dask-scheduler
    ```
- Run `dask-cuda-worker` on each worker node, providing the scheduler node's address. `dask-cuda-worker` will detect available GPUs and fork, starting one dask worker process per GPU.

    ```
    $ dask-cuda-worker 127.0.0.1:8786
    Start worker at:   tcp://127.0.0.1:36124
    Registered to: tcp://127.0.0.1:8786
    ```
- Connect to the cluster from a Python script
    ```
    >>> from dask.distributed import Client
    >>> # wrap Dask client API use in a main check
    >>> if __name__ == "__main__":
    ...     client = Client(address="127.0.0.1:8786")
    ...     # use_dask_APIs()
    ```
    Notes:
    - For more information on Dask CLI options use `dask-scheduler --help` and `dask-cuda-worker --help`

    More information on deploying manually can be found [here](https://docs.dask.org/en/latest/setup/cli.html).

>     Dask also supports other options like [SSH](https://docs.dask.org/en/latest/setup/ssh.html) and [NFS](https://docs.dask.org/en/latest/setup/hpc.html#using-a-shared-network-file-system-and-a-job-scheduler).

Reviewer: ```suggestion
Dask also supports other options like [SSH](https://docs.dask.org/en/latest/setup/ssh.html) and [NFS](https://docs.dask.org/en/latest/setup/hpc.html#using-a-shared-network-file-system-and-a-job-scheduler):
```
# starts one worker process per GPU on the supplied host list
dask-ssh 10.0.0.1 10.0.0.2 --remote-dask-worker dask_cuda.dask_cuda_worker
```
```

Reviewer: The suggestion formatting got messed up, but you should include basic examples of SSH and NFS dask-cuda cluster creation.

    ```
{: .mb-5}

#### **Kubernetes and Helm**
RAPIDS libraries can also be deployed using [Kubernetes](https://kubernetes.io) and [Helm](https://helm.sh). This is useful for deploying RAPIDS on existing Kubernetes clusters, or to managed cloud Kubernetes services.
- [Install Helm](https://helm.sh/docs/using_helm/#installing-helm)
- [Link-to-k8-script-and-charts]
- ```
    $ helm install [link-to-chart] --name rapids-dask --namespace rapids
  ```
- Check the status of the running pods
  ```
  $ kubectl get pods -w --namespace rapids
  ```
  Additional Information:
GPU scheduling in Kubernetes](https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/) 
[Dask Kubernetes](https://kubernetes.dask.org/en/latest/)
{: .mb-5}

#### **Docker**

##### **Docker Compose**
- Get the docker yml file [link-to-repo]
- ```
  $ docker compose ...
  ```

##### **Docker Swarm**

#### **Cloud deployments**

You can use a single type of configuration to deploy RAPIDS across multiple cloud providers with Kubernetes and Helm. You can also deploy RAPIDS through other managed service offerings where available.

> ##### **Google Dataproc**

Reviewer: Let's not include specific cloud provider services yet
